From 7da195c64bada14b7a5c8cce79488dfc52f5ecd5 Mon Sep 17 00:00:00 2001
From: Tomas Henzl <thenzl@redhat.com>
Date: Thu, 16 Oct 2014 16:02:03 -0400
Subject: [scsi] mpt3sas: delay scsi_add_host call to work with scsi-mq

Message-id: <1413475335-14737-21-git-send-email-thenzl@redhat.com>
Patchwork-id: 97600
O-Subject: [RHEL7.1 e-stor PATCH 20/32] mpt3sas: delay scsi_add_host call to work with scsi-mq
Bugzilla: 1088524
RH-Acked-by: Maurizio Lombardi <mlombard@redhat.com>
RH-Acked-by: Rob Evers <revers@redhat.com>

In _scsih_probe, delay the call to scsi_add_host until the host has been
fully set up.

Otherwise, the default .can_queue value of 1 causes scsi-mq to set the block
layer request queue size to its minimum size, resulting in awful performance.

In _scsih_probe error handling, call mpt3sas_base_detach rather than
scsi_remove_host to properly clean up in reverse order.

In _scsih_remove, call scsi_remove_host earlier to clean up in reverse order.

Signed-off-by: Robert Elliott <elliott@hp.com>
Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Nagalakshmi Nandigama <Nagalakshmi.Nandigama@avagotech.com>
Signed-off-by: Sreekanth Reddy <Sreekanth.Reddy@avagotech.com>
Signed-off-by: Christoph Hellwig <hch@lst.de>
(cherry picked from commit 4dc06fd84645c323a1f20482b9b571cab6dc7d93)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/mpt3sas/mpt3sas_base.c b/drivers/scsi/mpt3sas/mpt3sas_base.c
index c608a61..93ce2b2 100644
--- a/drivers/scsi/mpt3sas/mpt3sas_base.c
+++ b/drivers/scsi/mpt3sas/mpt3sas_base.c
@@ -266,7 +266,7 @@ mpt3sas_base_stop_watchdog(struct MPT3SAS_ADAPTER *ioc)
  ioc->fault_reset_work_q = NULL;
  spin_unlock_irqrestore(&ioc->ioc_reset_in_progress_lock, flags);
  if (wq) {
-  if (!cancel_delayed_work(&ioc->fault_reset_work))
+  if (!cancel_delayed_work_sync(&ioc->fault_reset_work))
    flush_workqueue(wq);
   destroy_workqueue(wq);
  }
diff --git a/drivers/scsi/mpt3sas/mpt3sas_scsih.c b/drivers/scsi/mpt3sas/mpt3sas_scsih.c
index a14be8f..9e0f824 100644
--- a/drivers/scsi/mpt3sas/mpt3sas_scsih.c
+++ b/drivers/scsi/mpt3sas/mpt3sas_scsih.c
@@ -2581,11 +2581,10 @@ _scsih_fw_event_cleanup_queue(struct MPT3SAS_ADAPTER *ioc)
   return;
 
  list_for_each_entry_safe(fw_event, next, &ioc->fw_event_list, list) {
-  if (cancel_delayed_work(&fw_event->delayed_work)) {
+  if (cancel_delayed_work_sync(&fw_event->delayed_work)) {
    _scsih_fw_event_free(ioc, fw_event);
    continue;
   }
-  fw_event->cancel_pending_work = 1;
  }
 }
 
@@ -7039,7 +7038,7 @@ static void
 _mpt3sas_fw_work(struct MPT3SAS_ADAPTER *ioc, struct fw_event_work *fw_event)
 {
  /* the queue is being flushed so ignore this event */
- if (ioc->remove_host || fw_event->cancel_pending_work ||
+ if (ioc->remove_host ||
      ioc->pci_error_recovery) {
   _scsih_fw_event_free(ioc, fw_event);
   return;
@@ -7439,9 +7438,9 @@ static void _scsih_remove(struct pci_dev *pdev)
  }
 
  sas_remove_host(shost);
+ scsi_remove_host(shost);
  mpt3sas_base_detach(ioc);
  list_del(&ioc->list);
- scsi_remove_host(shost);
  scsi_host_put(shost);
 }
 
@@ -7809,13 +7808,6 @@ _scsih_probe(struct pci_dev *pdev, const struct pci_device_id *id)
   }
  }
 
- if ((scsi_add_host(shost, &pdev->dev))) {
-  pr_err(MPT3SAS_FMT "failure at %s:%d/%s()!\n",
-      ioc->name, __FILE__, __LINE__, __func__);
-  list_del(&ioc->list);
-  goto out_add_shost_fail;
- }
-
  /* register EEDP capabilities with SCSI layer */
  if (prot_mask > 0)
   scsi_host_set_prot(shost, prot_mask);
@@ -7843,15 +7835,21 @@ _scsih_probe(struct pci_dev *pdev, const struct pci_device_id *id)
       ioc->name, __FILE__, __LINE__, __func__);
   goto out_attach_fail;
  }
+ if ((scsi_add_host(shost, &pdev->dev))) {
+  pr_err(MPT3SAS_FMT "failure at %s:%d/%s()!\n",
+      ioc->name, __FILE__, __LINE__, __func__);
+  list_del(&ioc->list);
+  goto out_add_shost_fail;
+ }
+
  scsi_scan_host(shost);
  return 0;
-
+out_add_shost_fail:
+ mpt3sas_base_detach(ioc);
  out_attach_fail:
  destroy_workqueue(ioc->firmware_event_thread);
  out_thread_fail:
  list_del(&ioc->list);
- scsi_remove_host(shost);
- out_add_shost_fail:
  scsi_host_put(shost);
  return -ENODEV;
 }
-- 
1.7.1